src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,348 @@
+use nalgebra::{RealField, Scalar};
+
+use crate::aabb::Bounded;
+use crate::bvh::{Bvh, BvhNode};
+use crate::ray::Ray;
+
+#[derive(Debug, Clone, Copy)]
+enum RestChild {
+    Left,
+    Right,
+    None,
+}
+
+/// Iterator to traverse a [`Bvh`] in order from nearest [`Aabb`] to farthest for [`Ray`],
+/// without memory allocations
+#[allow(clippy::upper_case_acronyms)]
+pub struct DistanceTraverseIterator<
+    'bvh,
+    'shape,
+    T: Scalar + Copy,
+    const D: usize,
+    Shape: Bounded<T, D>,
+    const ASCENDING: bool,
+> {
+    /// Reference to the Bvh to traverse
+    bvh: &'bvh Bvh<T, D>,
+    /// Reference to the input ray
+    ray: &'bvh Ray<T, D>,
+    /// Reference to the input shapes array
+    shapes: &'shape [Shape],
+    /// Traversal stack. 4 billion items seems enough?
+    stack: [(usize, RestChild); 32],
+    /// Position of the iterator in bvh.nodes
+    node_index: usize,
+    /// Size of the traversal stack
+    stack_size: usize,
+    /// Whether or not we have a valid node (or leaf)
+    has_node: bool,
+}
+
+impl<'bvh, 'shape, T, const D: usize, Shape: Bounded<T, D>, const ASCENDING: bool>
+    DistanceTraverseIterator<'bvh, 'shape, T, D, Shape, ASCENDING>
+where
+    T: Scalar + Copy + Scalar + RealField,
+{
+    /// Creates a new [`DistanceTraverseIterator `]
+    pub fn new(bvh: &'bvh Bvh<T, D>, ray: &'bvh Ray<T, D>, shapes: &'shape [Shape]) -> Self {
+        DistanceTraverseIterator {
+            bvh,
+            ray,
+            shapes,
+            stack: [(0, RestChild::None); 32],
+            node_index: 0,
+            stack_size: 0,
+            has_node: true,
+        }
+    }
+
+    /// Test if stack is empty.
+    fn is_stack_empty(&self) -> bool {
+        self.stack_size == 0
+    }
+
+    /// Push node onto stack.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `stack[stack_size]` is out of bounds.
+    fn stack_push(&mut self, nodes: (usize, RestChild)) {
+        self.stack[self.stack_size] = nodes;
+        self.stack_size += 1;
+    }
+
+    /// Pop the stack and return the node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `stack_size` underflows.
+    fn stack_pop(&mut self) -> (usize, RestChild) {
+        self.stack_size -= 1;
+        self.stack[self.stack_size]
+    }
+
+    /// Attempt to move to the nearest to ray node child of the current node.
+    /// If it is a leaf, or the [`Ray`] does not intersect the any node [`Aabb`], `has_node` will become false.
```

Comment:
I don't _really_ know what this is supposed to say. However as it stands, the grammar could be improved I think.
```suggestion
    /// Attempt to move to the child node that is closest to the ray, relative to the current node.
    /// If it is a leaf, or the [`Ray`] does not intersect the any node [`Aabb`], `has_node` will become `false`.
```

Reply:
I'm weak in english